src/library/scala/collection/immutable/Map.scala

```diff
@@ -410,3 +417,47 @@ object Map extends MapFactory[Map] {
 /** Explicit instantiation of the `Map` trait to reduce class file size in subclasses. */
 @SerialVersionUID(3L)
 abstract class AbstractMap[K, +V] extends scala.collection.AbstractMap[K, V] with Map[K, V]
+
+private[collection] final class MapBuilderImpl[K, V] extends Builder[(K, V), Map[K, V]] {
+  private[this] var elems: Map[K, V] = Map.empty
+  private[this] var hashMapBuilder: HashMapBuilder[K, V] = _
+
+  override def clear(): Unit = {
+    elems = Map.empty
+    if (hashMapBuilder != null) {
+      hashMapBuilder.clear()
+    }
+  }
+
+  override def result(): Map[K, V] = {
+    if (hashMapBuilder == null || hashMapBuilder.size == 0) {
+      elems
+    } else if (elems.size == 4) {
+      val map4 = elems.asInstanceOf[Map4[K, V]]
+
+      hashMapBuilder.addOneIfNotExists(map4.key1, map4.value1)
+      hashMapBuilder.addOneIfNotExists(map4.key2, map4.value2)
+      hashMapBuilder.addOneIfNotExists(map4.key3, map4.value3)
+      hashMapBuilder.addOneIfNotExists(map4.key4, map4.value4)
+
+      hashMapBuilder.result()
+    } else {
+      // should never happen...
+      elems.foreach { case (k, v) => hashMapBuilder.addOneIfNotExists(k, v) }
+      hashMapBuilder.result()
+    }
+  }
+  override def addOne(elem: (K, V)) = {
+    if (elems.size < 4) {
+      elems = elems + elem
+    } else {
+      if (hashMapBuilder == null) {
+        hashMapBuilder = new HashMapBuilder
```

Comment:
I would have immediately added `elems` to this `hashMapBuilder`, rather than doing that later in `result`. What's the motivation for the current organization?

Reply:
@retronym I was thinking that I may save on adding some of the 4 `Map4` elements, if some of those elements have since been added to the `HashMapBuilder`. But I didn't benchmark this so have no evidence it actually performs better so we can revert to your simpler implementation :-)

Reply:
Fodder for a followup PR:

  - we could override `addAll` here to go directly to `hashMapBuilder` if input collection is a map with > 4 elements
  - or maybe, just do all the building into a `hashMapBuilder`, and add a special case in `result` to directly create a `MapN` of the right size when the build hash map is small. We'd then have the space advantage of the small, built maps, and building performance might even require less allocations to go through `Map1` => `Map2` etc.